epistasis_selection.py

```diff
@@ -0,0 +1,98 @@
+  {M \choose K} * 2^{P} variants. See the definition for `utils.merge_mutation_sets` for more on
+  mutation merging.
+
+  Args:
+    num_rounds: The number of rounds of combination
+    mutations: The starting pool of mutations, where each mutation is an iterable of
+      tuples encoding mutations (position, mutation).
+
+  Returns:
+    A list of tuples of mutations, where each element will be a combination of
+    `num_rounds` mutations from `mutations`. Note that each tuple will possibly be of different lengths.
+
+  """
+  if num_rounds == 0:
+    return mutations
+  mutations_to_combine = list(itertools.combinations(mutations, num_rounds + 1))
+
+  all_samples = []
+
+  for mutant in mutations_to_combine:
+    mutant = list(mutant)
+    prev_round = [()]
+    # iteratively add the mutations together
+    for i in range(num_rounds + 1):
+      next_round = []
+      new_mutation = mutant.pop()
+      for merged in prev_round:
+        next_round.extend(utils.merge_mutation_sets(merged, new_mutation))
+      prev_round = next_round
+    all_samples.extend(next_round)
+  return all_samples
+
+
+def get_test_seqs_for_landscape(landscape: potts_model.PottsModel,
+                                distance: int,
+                                n: int,
+                                adaptive: bool = True,
+                                top_k: Optional[int] = None,
+                                rng: np.random.Generator = np.random.default_rng(0)):
+  """Sample n epistatic variants from distance
+
+  Args:
+    landscape: the landscape in question
+    distance: the distance desired for the test set
+    n: the number of variants in the test set
+    adaptive: a boolean indicating whether you want the set to be enriched for adaptive or deleterious epistasis
+    rng: a random state. required so that there is no default randomness...
+
+  Return:
+    sequences for the test set
+  """
+  if distance % 2 != 0:
+    raise ValueError('Odd distance not supported.')
+
+  # take enough so that you can get unique variants... but not too many that you blow up
+  # another option is to split into a few rounds. Do one, see if you get it, try again, etc.
+  if not top_k:
+    top_k = n
+  tensor_indexes = utils.get_top_n_4d_tensor_indexes(landscape.epistasis_tensor, top_k, lowest=not adaptive)
+  mutation_pairs = [utils.get_mutation_pair_from_tensor_index(
+      t) for t in tensor_indexes]
+
+
+  num_rounds = distance // 2
+  all_combined = combine_k_rounds(num_rounds, mutation_pairs)
+  all_combined = _filter_elements_to_length(all_combined, distance)
+  if len(all_combined) < n:
+    raise ValueError(f'Not enough ({len(all_combined)} < {n}) mutants at distance {distance}, try increasing `top_k`.')
+  subset = rng.choice(all_combined, n, replace=False)
+  test_seqs = [utils.apply_mutations(landscape.wildtype_sequence, m) for m in subset]
+  return test_seqs
+
+
+def _filter_elements_to_length(elements, length):
```

Comment:
why can't you just do this?
return [element for element in elements if len(element) == length]

Reply:
definitely let's do that haha. not sure why I wanted this more complicated filtering mechanism